### PR TITLE
[WIP] Upgrade testem & ember-cli

### PR DESCRIPTION
**Pull request by jacobq**

### CODE REVIEW

File: lib/utils/test-task-mixin.js [hunk 1]
```diff
@@ -21,7 +21,7 @@ module.exports = {
       testemOptions.file = 'testem-electron.js';
     }
     else {
-      deprecate("In ember-cli v3.2.0 and newer, testemOptions is now called defaultOptions. Please consider upgrading to use the new method.", false);
+      deprecate("In ember-cli v3.2.0 and newer, testemOptions is now called defaultOptions. Please consider upgrading to use the new method.", true);
```
Comment (bendemboski) [inline lib/utils/test-task-mixin.js]: I don't think you need this deprecation warning here.

If the host application is using `ember-cli < v3.2`, then `this._super(...arguments)` will return a `testemOptions` object and everything will work fine (and the else clause won't be executed). If the host application is using `ember-cli >= 3.2` then this method won't be called at all, the else clause (and the rest of the function) won't be executed.

You're upgrading code inside the addon to be compatible with both old and new `ember-cli`, so there's nothing to deprecate...

Comment (jacobq) [reply lib/utils/test-task-mixin.js]: I guess I thought consumers might be calling `testemOptions`, in which case they'd want to migrate to the new API at some point. I'm probably over-thinking this though. If it's just internal then you're right, we should probably just use if-else/try-both logic in there and go on our merry way.

**Comment by felixrieseberg** [comment]
Sometimes I wonder if I should go and write a giant Electron mock. At Slack, we, for instance, do our unit tests in Jest now and just mock all of Electron – that way we don't have to worry about running the tests _actually_ inside Electron.

It'd be quite a bit of work and a major breaking change, but I'm wondering if this will just continue to keep breaking us here...

**Comment by jacobq** [comment]
@felixrieseberg While I follow your reasoning and share your sentiment, I think there's a lot of value in the  "test what you fly and fly what you test" philosophy. Yes, test things outside of electron that don't need electron, but mocking out a huge API like electron, aside from being a lot of work, decreases the value of tests. For example, my app was affected by https://github.com/electron/electron/issues/12029 but thankfully the problem was caught this in testing. It would've gone unnoticed longer if the tests had just used a mocked API (that didn't also mimic this particular bug).

**Comment by jacobq** [comment]
My nasty mucking about seems to have at least yield a good clue: module resolution is failing because in the context of the `require` statement, the search is `ember-electron` (tarball's `package`) not `ee-test-app` like it should be. I am not yet sure why this is or why the `npm` version matters, but the good news is that this seems to be a problem with the test rather than an actual problem that would affect users.

<summary><details>

```
JRQ-DEBUG: (index.js) node_modules path --> /tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules
JRQ-DEBUG: (index.js) ember-cli files.length --> 20
JRQ-DEBUG: (index.js) module globalPaths --> [ '/home/jquant/.node_modules',
  '/home/jquant/.node_libraries',
  '/home/jquant/.nvm/versions/node/v8.11.3/lib/node' ]
JRQ-DEBUG: (index.js) module _nodeModulePaths --> [ '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules',
  '/tmp/tmp-237907PhBrC23jDUl/node_modules',
  '/tmp/node_modules',
  '/node_modules' ]
JRQ-DEBUG: _resolveFilename called with ember-cli Module {
  id: '/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/ember-electron/index.js',
  exports: {},
  parent: 
   Module {
     id: '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/models/blueprint.js',
     exports: 
      { [Function: Class]
        lookup: [Function],
        load: [Function],
        list: [Function],
        _existsSync: [Function],
        _readdirSync: [Function],
        renamedFiles: [Object],
        ignoredFiles: [Array],
        ignoredUpdateFiles: [Array],
        defaultLookupPaths: [Function] },
     parent: 
      Module {
        id: '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/tasks/generate-from-blueprint.js',
        exports: [Function: GenerateTask],
        parent: [Object],
        filename: '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/tasks/generate-from-blueprint.js',
        loaded: true,
        children: [Array],
        paths: [Array] },
     filename: '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/models/blueprint.js',
     loaded: true,
     children: 
      [ [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Circular] ],
     paths: 
      [ '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/models/node_modules',
        '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/node_modules',
        '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/node_modules',
        '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules',
        '/tmp/tmp-237907PhBrC23jDUl/node_modules',
        '/tmp/node_modules',
        '/node_modules' ] },
  filename: '/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/ember-electron/index.js',
  loaded: false,
  children: 
   [ Module {
       id: '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules/fs-extra/lib/index.js',
       exports: [Object],
       parent: [Circular],
       filename: '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules/fs-extra/lib/index.js',
       loaded: true,
       children: [Array],
       paths: [Array] },
     Module {
       id: '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules/rsvp/dist/rsvp.js',
       exports: [Object],
       parent: [Circular],
       filename: '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules/rsvp/dist/rsvp.js',
       loaded: true,
       children: [],
       paths: [Array] } ],
  paths: 
   [ '/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/ember-electron/node_modules',
     '/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/node_modules',
     '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules',
     '/tmp/tmp-2379049WhZVq0CbxJ/node_modules',
     '/tmp/node_modules',
     '/node_modules' ] } false
JRQ-DEBUG: (index.js) caught exception while trying to load blueprint model { Error: Cannot find module 'ember-cli'
    at Module._resolveFilename (module.js:547:15)
    at Function.Module._resolveFilename (/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/ember-electron/index.js:32:12)
    at Function.resolve (internal/module.js:18:19)
    at Object.<anonymous> (/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/ember-electron/index.js:35:62)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Module.require (module.js:596:17) code: 'MODULE_NOT_FOUND' }
JRQ-DEBUG: _resolveFilename called with ember-cli/lib/models/blueprint Module {
  id: '/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/ember-electron/index.js',
  exports: {},
  parent: 
   Module {
     id: '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/models/blueprint.js',
     exports: 
      { [Function: Class]
        lookup: [Function],
        load: [Function],
        list: [Function],
        _existsSync: [Function],
        _readdirSync: [Function],
        renamedFiles: [Object],
        ignoredFiles: [Array],
        ignoredUpdateFiles: [Array],
        defaultLookupPaths: [Function] },
     parent: 
      Module {
        id: '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/tasks/generate-from-blueprint.js',
        exports: [Function: GenerateTask],
        parent: [Object],
        filename: '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/tasks/generate-from-blueprint.js',
        loaded: true,
        children: [Array],
        paths: [Array] },
     filename: '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/models/blueprint.js',
     loaded: true,
     children: 
      [ [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Circular] ],
     paths: 
      [ '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/models/node_modules',
        '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/lib/node_modules',
        '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules/ember-cli/node_modules',
        '/tmp/tmp-237907PhBrC23jDUl/ee-test-app/node_modules',
        '/tmp/tmp-237907PhBrC23jDUl/node_modules',
        '/tmp/node_modules',
        '/node_modules' ] },
  filename: '/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/ember-electron/index.js',
  loaded: false,
  children: 
   [ Module {
       id: '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules/fs-extra/lib/index.js',
       exports: [Object],
       parent: [Circular],
       filename: '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules/fs-extra/lib/index.js',
       loaded: true,
       children: [Array],
       paths: [Array] },
     Module {
       id: '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules/rsvp/dist/rsvp.js',
       exports: [Object],
       parent: [Circular],
       filename: '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules/rsvp/dist/rsvp.js',
       loaded: true,
       children: [],
       paths: [Array] } ],
  paths: 
   [ '/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/ember-electron/node_modules',
     '/tmp/tmp-2379049WhZVq0CbxJ/package/blueprints/node_modules',
     '/tmp/tmp-2379049WhZVq0CbxJ/package/node_modules',
     '/tmp/tmp-2379049WhZVq0CbxJ/node_modules',
     '/tmp/node_modules',
     '/node_modules' ] } false
```
</details></summary>

**Comment by jacobq** [comment]
Thanks @stopdropandrew for the npm work-around here https://github.com/jacobq/ember-electron/pull/2
Rather than try to merge and squash here, I'm just going to start a fresh PR to keep it clean -- redoing or cherry-picking as appropriate.

**Comment by jacobq** [comment]
Closing in favor of https://github.com/felixrieseberg/ember-electron/pull/355

**Pull request closed without merging**